snuba/datasets/entity.py

```diff
@@ -156,3 +174,129 @@ def describe(self) -> Description:
                 Description(header="Relationships", content=relationships),
             ],
         )
+
+
+class InvalidSubscriptionError(Exception):
+    pass
+
+
+class EntitySubscription:
+    """
+    Used to provide extra functionality for entity specific subscriptions.
+
+    An instance of this class is set as an attribute of Entity.
+        Entity.entity_subscription: Optional[EntitySubscription]
+    """
+
+    def __init__(
+        self,
+        max_allowed_aggregations: Optional[int] = None,
+        disallowed_aggregations: Optional[Sequence[str]] = None,
+        organization: Optional[int] = None,
+    ) -> None:
+        max_allowed_aggregations = max_allowed_aggregations
+        disallowed_aggregations = disallowed_aggregations
+        organization = organization
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        """
+        Applies entity specific validations on query argument passed
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> EntitySubscription:
+        raise NotImplementedError
+
+
+class EntitySubscriptionValidation:
+    def __init__(
+        self,
+        max_allowed_aggregations: int = 1,
+        disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"],
+        organization: Optional[int] = None,
+    ) -> None:
+        self.max_allowed_aggregations = max_allowed_aggregations
+        self.disallowed_aggregations = disallowed_aggregations
+        self.organization = organization
+
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        # Import get_entity() when used, not at import time to avoid circular imports
+        from snuba.datasets.entities.factory import get_entity
+
+        # TODO: Support composite queries with multiple entities.
+        from_clause = query.get_from_clause()
+        if not isinstance(from_clause, EntityDS):
+            raise InvalidSubscriptionError("Only simple queries are supported")
+        entity = get_entity(from_clause.key)
+
+        SubscriptionAllowedClausesValidator(
+            self.max_allowed_aggregations, self.disallowed_aggregations
+        ).validate(query)
+        if entity.required_time_column:
+            NoTimeBasedConditionValidator(entity.required_time_column).validate(query)
+
+
+class BaseEntitySubscription(EntitySubscriptionValidation, EntitySubscription):
```

Comment:
I think you should rename this class to `EntitySubscription` and collapse the other two classes into it. We don't need to extend this, and having the validation be separate isn't useful.

I also think all this code shouldn't be in the entity file. I am actually OK with it being in the `entity_subscription` folder, unless that causes import cycles.

Reply:
Collapsed classes and renamed.